lib/src/components/NxCombobox/NxCombobox.tsx

```diff
@@ -0,0 +1,311 @@
+        [alertDropdownId]: showAlert
+      });
+
+  // There is a requirement that when there is an error querying the data, if the user navigates away from
+  // the component and then comes back to it the search should be retried automatically
+  function handleComponentFocus(evt: FocusEvent<HTMLDivElement>) {
+
+    if (loadError) {
+      // check if this is focus coming into the component from somewhere else on the page, not just moving between
+      // children of this component and not from focus coming into the browser from some other window
+      const comingFromOutsidePage = evt.relatedTarget === null,
+          comingFromChildNode = evt.relatedTarget instanceof Node && evt.currentTarget.contains(evt.relatedTarget);
+
+      if (!(comingFromOutsidePage || comingFromChildNode)) {
+        doSearch(value);
+      }
+    }
+  }
+
+  function handleComponentBlur(evt: FocusEvent<HTMLDivElement>) {
+
+    if (!(evt.relatedTarget instanceof Node && evt.currentTarget.contains(evt.relatedTarget))) {
+      // The automatically selected suggestion becomes the value of the combobox
+      // when the combobox loses focus.
+      if (autoComplete && focusableBtnIndex !== null) {
+        const elToFocusText = matches[focusableBtnIndex].displayName;
+        onChange(elToFocusText);
+        onSearch(elToFocusText.trim());
+      }
+      setFocusableBtnIndex(null);
+      setInlineStyle(false);
+    }
+  }
+
+  function getDropdownBtnIdForIndex(idx: number) {
+    return `${dropdownBtnIdPrefix}-${idx}`;
+  }
+
+  function handleOnChange(newVal: string) {
+    setFocusableBtnIndex(null);
+    onChange(newVal);
+
+    if (newVal.trim() !== value.trim()) {
+      doSearch(newVal);
+    }
+  }
+
+  function doSearch(val: string) {
+    focusTextInput();
+    onSearch(val.trim());
+  }
+
+  function focusTextInput() {
+    inputRef.current?.querySelector('input')?.focus();
+  }
+
+  // helper for focusing different buttons in the dropdown menu
+  const adjustBtnFocus = (adjust: (i: number) => number) => () => {
+        const newFocusableBtnIndex = adjust(focusableBtnIndex ?? 0),
+            elToFocus = dropdownRef.current?.children[newFocusableBtnIndex] as HTMLElement | null;
+
+        if (elToFocus) {
+          setFocusableBtnIndex(newFocusableBtnIndex);
+          elToFocus.scrollIntoView({ block: 'nearest' });
+          setInlineStyle(false);
+        }
+      },
+      focusNext = adjustBtnFocus(inc),
+      focusPrev = adjustBtnFocus(dec),
+      focusFirst = adjustBtnFocus(always(0)),
+      focusLast = adjustBtnFocus(always(matches.length - 1));
+
+  function handleOnClick({displayName}: DataItem<T, string>) {
+    onChange(displayName);
+    onSearch(displayName.trim());
+    focusTextInput();
+    setFocusableBtnIndex(null);
+    setInlineStyle(false);
+  }
+
+  function handleKeyUp(evt: KeyboardEvent<HTMLElement>) {
+    if (autoComplete) {
+      const isPrintableChar = evt.key.length === 1 && evt.key.match(/[!-~]/);
```

Comment:
This regex only matches non-whitespace ASCII characters.  It does not match non-ASCII printable unicode characters.  I also wonder if assuming that the autocomplete should be removed in the event of any non-printable character is dangerously general.  I imagine there's only a handful of keys that we're specifically picturing this capturing - arrows, ESC, Home/End, maybe a few others.  Should we just handle them specifically.  I'm also wondering why this needs to be done on keyup, could it be added to the keydown handler instead?

Reply:
My intension here is to capture key combinations(e.g. on mac, press `command` key + `a` key to select all text in the input box) which should not have `setShowAutoComplete` set to true. 
I had it in the keydown handler before but seems like it could not check the key combinations, when pressing `command` key + `a` key, once the `a` key is pressed down, the keydown handler is triggered even though the `command` key is still pressed.

Reply:
Hi Ross, thanks for the helps! I noticed some behaviors and wasn't sure if they are expected.
1) when pressing `backspace` key, if there are matches, should it still show autoComplete string? I thought we want to keep the text after pressing 'backspace' key.
2) when navigating in the dropdown with arrow keys, should we still keep the `selectionRange`?
3) In the predetermined lists example(`autoComplete=true`), type "a" and wait until "Alabama" showed, then press `escape` key, it will clear "Alabama" first then immediately show the name of the first item in the lists, I think we want to keep it empty.

Reply:
> I wouldn't think so; does it currently? When you navigate the dropdown with arrow keys, I'd expect that to fire an onChange event with the full text of each dropdown option, which would typically cause that option to get fed back in as the value prop, causing the value to become equal to the full inputVal

Oh I see, because we don't call `onChange` until after blur.  I will work on this

Reply:
> My assumption was that we'd want it to behave the way a text input with a partial selection usually behaves when Backspace is pressed: by deleting the highlighted portion and keeping the rest. And in our case, unselecting the menu item.

Further on this one, after the highlighted portion is removed, then pressing `Backspace` key again, it will start searching, once there are matches, it showed autoComplete string with highlighted portion in the input box, do we want this as well?

Reply:
> Further on this one, after the highlighted portion is removed, then pressing Backspace key again, it will start searching, once there are matches, it showed autoComplete string with highlighted portion in the input box, do we want this as well?

See the latest on my branch.  I _think_ I have everything worked out at this point.